app/src/controller.c

```diff
@@ -45,9 +73,67 @@ controller_destroy(struct controller *controller) {
     receiver_destroy(&controller->receiver);
 }
 
+static void
+control_msg_log(const struct control_msg *msg) {
+    switch(msg->type) {
+        case CONTROL_MSG_TYPE_INJECT_KEYCODE:
+            LOGV("Event: key %s code=%u repeat=%u meta=%06x",
```

Comment:
The actual type of an enum is implementation defined ([specs](https://port70.net/~nsz/c/c11/n1570.html#6.7.2.2p4)), so the value should be cast to a concrete integer type (`int` for example, with `"%d"` as format),

The `repeat` value is `uint32_t`, so the format must be `PRIu32`.

_Idem for other values._

```c
LOGV("Event: key %s code=%d repeat=%" PRIu32 " meta=%06lx",
    ENUM_TO_DESCRIPTION(android_keyevent_action_descriptions, msg->inject_keycode.action),
    (int) msg->inject_keycode.keycode,
    msg->inject_keycode.repeat,
    (long) msg->inject_keycode.metastate); // in theory, int may be as small as 16bits, so not sufficient
```
_(keep your own indentation, I just wrote it from the github interface.)_